sky/backends/backend_utils.py

```diff
@@ -886,6 +898,16 @@ def write_cluster_config(
     yaml_config = common_utils.read_yaml(tmp_yaml_path)
     config_dict['cluster_name_on_cloud'] = yaml_config['cluster_name']
 
+    # Make sure to do this before we optimize file mounts. Optimization is
+    # non-deterministic, but everything else before this point should be
+    # deterministic.
+    try:
+        config_dict['config_hash'] = _deterministic_cluster_yaml_hash(
+            tmp_yaml_path)
+    except Exception as e:  # pylint: disable=broad-except
+        logger.warning(f'Failed to calculate config_hash: {e}')
```

Comment:
We can keep this in `debug` as a user may not need to know the details.

Reply:
See above comment https://github.com/skypilot-org/skypilot/pull/4289#discussion_r1868307331. I think this is pretty bad and a warning is appropriate.

Reply:
I see, if the chance for the exception is very low, it should be fine. 

nit: use `common_utils.format_exception(e)` to replace `e`?